esp-wifi/src/wifi/mod.rs

```diff
@@ -3193,247 +2713,585 @@ pub(crate) fn apply_power_saving(ps: PowerSaveMode) -> Result<(), WifiError> {
+            include::esp_wifi_clear_ap_list();
         }
+    }
+}
 
-        /// An async wifi network scan with caller-provided scanning options.
-        pub async fn scan_with_config_async<const N: usize>(
-            &mut self,
-            config: ScanConfig<'_>,
-        ) -> Result<(heapless::Vec<AccessPointInfo, N>, usize), WifiError> {
-            Self::clear_events(WifiEvent::ScanDone);
-            esp_wifi_result!(wifi_start_scan(false, config))?;
+#[non_exhaustive]
+pub struct Interfaces<'d> {
+    pub sta: WifiDevice<'d>,
+    pub ap: WifiDevice<'d>,
+}
 
-            // Prevents memory leak if `scan_n`'s future is dropped.
-            let guard = FreeApListOnDrop;
-            WifiEventFuture::new(WifiEvent::ScanDone).await;
+#[cfg(feature = "smoltcp")]
+fn timestamp() -> smoltcp::time::Instant {
+    smoltcp::time::Instant::from_micros(
+        esp_hal::time::Instant::now()
+            .duration_since_epoch()
+            .as_micros() as i64,
+    )
+}
 
-            guard.defuse();
+/// Create a WiFi controller and it's associated interfaces.
+///
+/// Dropping the controller will deinitialize / stop WiFi.
+pub fn new<'d>(
+    inited: &'d EspWifiController<'d>,
+    _device: impl crate::hal::peripheral::Peripheral<P = crate::hal::peripherals::WIFI> + 'd,
+) -> Result<(WifiController<'d>, Interfaces<'d>), WifiError> {
+    if !inited.wifi() {
+        crate::wifi::wifi_init()?;
+    }
+    Ok((
+        WifiController {
+            _phantom: Default::default(),
+            #[cfg(feature = "sniffer")]
+            sniffer_taken: AtomicBool::new(false),
+        },
+        Interfaces {
+            sta: WifiDevice {
+                _phantom: Default::default(),
+                mode: WifiDeviceMode::Sta,
+            },
+            ap: WifiDevice {
+                _phantom: Default::default(),
+                mode: WifiDeviceMode::Ap,
+            },
+        },
+    ))
+}
 
-            let count = self.scan_result_count()?;
-            let result = self.scan_results()?;
+#[non_exhaustive]
+pub struct WifiController<'d> {
+    _phantom: PhantomData<&'d ()>,
+    #[cfg(feature = "sniffer")]
+    sniffer_taken: AtomicBool,
+}
+
+impl Drop for WifiController<'_> {
+    fn drop(&mut self) {
+        if unwrap!(
+            crate::flags::WIFI.fetch_update(Ordering::SeqCst, Ordering::SeqCst, |x| {
+                Some(x.saturating_sub(1))
+            })
+        ) == 0
+        {
+            if let Err(e) = crate::wifi::wifi_deinit() {
+                warn!("Failed to cleanly deinit wifi: {:?}", e);
+            }
+        }
+    }
+}
 
-            Ok((result, count))
+impl WifiController<'_> {
+    /// Attempts to take the sniffer, returns `Some(Sniffer)` if successful,
+    /// otherwise `None`.
+    #[cfg(feature = "sniffer")]
+    pub fn take_sniffer(&self) -> Option<Sniffer> {
+        if self
+            .sniffer_taken
+            .compare_exchange(false, true, Ordering::Acquire, Ordering::Relaxed)
```

Comment:
This can be a `fetch_or`.

Reply:
using fetch_or but honestly that sniffer and esp-now should become (very) special interfaces (and the sniffer API could use some love ... like tx takes two booleans where I have no idea what is what unless looking at the docs/code)

but given this PR is already not fun to look at let save that for later (I plan to create an issue with a few things to address once this landed)